pyomo/repn/plugins/gams_writer.py

```diff
@@ -92,27 +94,155 @@ def visiting_potential_leaf(self, node):
         if node.__class__ in native_types:
             return True, str(node)
 
+        if node.is_expression_type():
+            # we will descend into this, so type checking will happen later
+            if node.is_component_type():
+                self.treechecker(node)
+            return False, None
+
+        if node.is_component_type():
+            if self.ctype(node) not in valid_expr_ctypes_minlp:
+                # Make sure all components in active constraints
+                # are basic ctypes we know how to deal with.
+                raise RuntimeError(
+                    "Unallowable component '%s' of type %s found in an active "
+                    "constraint or objective.\nThe GAMS writer cannot export "
+                    "expressions with this component type."
+                    % (node.name, self.ctype(node).__name__))
+            if self.ctype(node) is not Var:
+                # For these, make sure it's on the right model. We can check
+                # Vars later since they don't disappear from the expressions
+                self.treechecker(node)
+
         if node.is_variable_type():
             if node.fixed:
                 return True, str(value(node))
             label = self.smap.getSymbol(node)
             return True, label
 
-        if not node.is_expression_type():
-            return True, str(value(node))
+        return True, str(value(node))
 
-        return False, None
+    def ctype(self, comp):
+        if isinstance(comp, IComponent):
+            return comp.ctype
+        else:
+            return comp.type()
```

Comment:
What is the point of the `ctype()` helper function?  What case are you catching here?

Reply:
For kernel components, it's a property called `ctype`. For regular components, it's a method called `type()`.

Reply:
OK.  @ghackebeil: why is there this inconsistency?  Should we move `core.base` to also define a `ctype` property?  Move both to have a `ctype()` method?  I for one never liked the move from a `ctype` attribute to a `type()` method because it doesn't return the `type`, but rather the effective component type.  This is highlighted by the developers referring to it as `ctype` in conversation.

Regardless, I think we are good to merge and should move this discussion to an Issue / subsequent PR.